File: packages/prism-core/src/layer1/manifest/manifest-types.ts
/**
 * Prism Manifest types — the on-disk definition file for a workspace.
 *
 * ## Glossary (from SPEC.md)
 *
 *   Vault       — Encrypted local directory. The physical security boundary.
 *                  Files appear as unreadable blobs unless the Prism Daemon is
 *                  unlocked. A vault contains Collections and Manifests.
 *
 *   Collection  — A typed CRDT array (e.g. `Contacts`, `Audio_Busses`, `Tasks`).
 *                  Collections hold the actual data. Multiple manifests can
 *                  reference the same collection.
 *
 *   Manifest    — A YAML/JSON file containing weak references to Collections.
 *                  A "workspace" in user-facing terms is just a Manifest —
 *                  it points to data nodes, it does not contain them.
 *                  Example: a "JJM Productions" manifest in Flux points to
 *                  Contacts, Projects, and Tasks collections. A personal and
 *                  a professional manifest can both point to the same Contacts
 *                  collection, filtering via tags.
 *
 *   Shell       — The IDE chrome that renders whatever a Manifest references.
 *                  The shell has no fixed layout; it derives content from
 *                  registries (LensRegistry, ViewRegistry).
 *
 * ## File format
 *
 * Stored as `.prism.json` at the root of a vault directory.
 *
 * Ported from @core/project in legacy Helm. Adapted for Prism:
 *   - Loro CRDT is the primary storage, not SQLite
 *   - Tauri IPC, not HTTP — no http storage backend
 *   - Simplified sync to peer-based CRDT replication
 *   - Helm→Prism rename throughout
 */

// ── Storage config ─────────────────────────────────────────────────────────────

export type StorageBackend = "loro" | "memory" | "fs";

/** Loro CRDT document storage (default). */
export interface LoroStorageConfig {
  backend: "loro";
  /** Path to the Loro document file, relative to vault root. */
  path: string;
}

/** In-memory storage (testing / ephemeral). */
export interface MemoryStorageConfig {
  backend: "memory";
}

/** Filesystem-backed storage (Tauri fs). */
export interface FsStorageConfig {
  backend: "fs";
  /** Directory path relative to vault root. */
  directory: string;
}

export type StorageConfig =
  | LoroStorageConfig
  | MemoryStorageConfig
  | FsStorageConfig;

// ── Schema config ──────────────────────────────────────────────────────────────

export interface SchemaConfig {
  /**
   * Schema module references, resolved in order.
   *   - Built-in module: '@prism/core'
   *   - Relative path: './schemas/custom.yaml'
   * @default ['@prism/core']
   */
  modules: string[];
}

// ── Sync config ────────────────────────────────────────────────────────────────

export type SyncMode = "off" | "manual" | "auto";

export interface SyncConfig {
  mode: SyncMode;
  /** Sync interval in seconds when mode is 'auto'. 0 = continuous. */
  intervalSeconds?: number | undefined;
  /** Peer addresses for CRDT sync. */
  peers?: string[] | undefined;
}

// ── Collection reference ──────────────────────────────────────────────────────

/**
 * A reference to a Collection from within a Manifest.
 *
 * Collections are typed CRDT arrays that hold the actual data.
 * A CollectionRef is the manifest's pointer to a collection,
 * optionally with filters to narrow what's visible in this context.
 */
export interface CollectionRef {
  /** Unique id of this reference within the manifest. */
  id: string;
  /** Display name for this collection in the shell. */
  name: string;
  description?: string | undefined;
  /** Object type filter — only objects of these types appear. Empty = all. */
  objectTypes?: string[] | undefined;
  /** Tag filter — objects must have all listed tags. */
  tags?: string[] | undefined;
  /** Sort field. */
  sortBy?: string | undefined;
  sortDirection?: ("asc" | "desc") | undefined;
  /** Icon identifier. */
  icon?: string | undefined;
}

// ── Manifest ────────────────────────────────────────────────────────────────

export const MANIFEST_FILENAME = ".prism.json";
export const MANIFEST_VERSION = "1";

export type ManifestVisibility = "private" | "team" | "public";

/**
 * A Prism Manifest — the on-disk definition of a "workspace".
 *
 * A manifest is a named set of weak references to Collections inside a Vault.
 * It defines *what data to show* and *how to configure the shell*, but it
 * does not hold data itself. Multiple manifests can reference the same
 * collections with different filters.
 */
export interface PrismManifest {
  /** Stable UUID — identifies this manifest uniquely. */
  id: string;
  /** Display name shown in the title bar and recent list. */
  name: string;
  /** Schema version of this manifest format. Currently '1'. */
  version: string;
  /** Storage backend configuration for the vault. */
  storage: StorageConfig;
  /** Schema module configuration. */
  schema: SchemaConfig;
  /** Optional sync configuration. */
  sync?: SyncConfig | undefined;
  /** Collection references — weak pointers to typed CRDT arrays. */
  collections?: CollectionRef[] | undefined;
  /** ISO timestamp when this manifest was created. */
  createdAt: string;
  /** ISO timestamp when this manifest was last opened. */
  lastOpenedAt?: string | undefined;

  // ── Modules & settings ───────────────────────────────────────────────────

  /** Plugin/module enable flags. Keys = plugin IDs, values = enabled. */
  modules?: Record<string, boolean> | undefined;
  /** Manifest-scoped settings (dot-notation keys). */
  settings?: Record<string, unknown> | undefined;

  // ── Ownership ────────────────────────────────────────────────────────────

  /** DID or user ID of the manifest owner. */
  ownerId?: string | undefined;
  /** Visibility level. Default: 'private'. */
  visibility?: ManifestVisibility | undefined;
  /** Human-readable description. */
  description?: string | undefined;
}

File: packages/prism-core/src/identity/manifest/manifest_types.rs
//! On-disk definition types for a Prism vault. Ported from
//! `identity/manifest/manifest-types.ts`.
//!
//! Glossary (from `SPEC.md`): a **Vault** is the encrypted directory,
//! a **Collection** is a typed CRDT array holding the actual data, a
//! **Manifest** is a weak-reference file pointing to collections, and
//! the **Shell** is the chrome that renders whatever the manifest
//! references.

use indexmap::IndexMap;
use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::privilege_set::{PrivilegeSet, RoleAssignment};

// ── Storage config ─────────────────────────────────────────────────

/// Tagged union of the three storage backends. JSON shape matches the
/// TS discriminated union on the `backend` key.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "backend", rename_all = "lowercase")]
pub enum StorageConfig {
    Loro {
        /// Path to the Loro document file, relative to vault root.
        path: String,
    },
    Memory,
    Fs {
        /// Directory path relative to vault root.
        directory: String,
    },
}

impl StorageConfig {
    /// Name of the active backend — `"loro"`, `"memory"`, or `"fs"`.
    /// Used by [`validate_manifest`] when we need to dump a backend
    /// name back into an error message.
    pub fn backend_name(&self) -> &'static str {
        match self {
            StorageConfig::Loro { .. } => "loro",
            StorageConfig::Memory => "memory",
            StorageConfig::Fs { .. } => "fs",
        }
    }
}

// ── Schema config ──────────────────────────────────────────────────

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SchemaConfig {
    pub modules: Vec<String>,
}

// ── Sync config ────────────────────────────────────────────────────

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SyncMode {
    Off,
    Manual,
    Auto,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SyncConfig {
    pub mode: SyncMode,
    #[serde(
        default,
        rename = "intervalSeconds",
        skip_serializing_if = "Option::is_none"
    )]
    pub interval_seconds: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub peers: Option<Vec<String>>,
}

// ── Collection reference ───────────────────────────────────────────

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SortDirection {
    Asc,
    Desc,
}

/// A reference to a Collection from within a Manifest. Weakly points
/// at the data, never contains it.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct CollectionRef {
    pub id: String,
    pub name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(
        default,
        rename = "objectTypes",
        skip_serializing_if = "Option::is_none"
    )]
    pub object_types: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
    #[serde(default, rename = "sortBy", skip_serializing_if = "Option::is_none")]
    pub sort_by: Option<String>,
    #[serde(
        default,
        rename = "sortDirection",
        skip_serializing_if = "Option::is_none"
    )]
    pub sort_direction: Option<SortDirection>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub icon: Option<String>,
}

impl CollectionRef {
    /// Convenience constructor used in tests and templates.
    pub fn new(id: impl Into<String>, name: impl Into<String>) -> Self {
        Self {
            id: id.into(),
            name: name.into(),
            description: None,
            object_types: None,
            tags: None,
            sort_by: None,
            sort_direction: None,
            icon: None,
        }
    }
}

// ── Manifest ───────────────────────────────────────────────────────

pub const MANIFEST_FILENAME: &str = ".prism.json";
pub const MANIFEST_VERSION: &str = "1";

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ManifestVisibility {
    Private,
    Team,
    Public,
}

/// A Prism Manifest — the on-disk definition of a workspace. Field
/// naming follows the TS JSON layout (camelCase) so on-disk files
/// round-trip through serde without a rename pass.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PrismManifest {
    pub id: String,
    pub name: String,
    pub version: String,
    pub storage: StorageConfig,
    pub schema: SchemaConfig,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sync: Option<SyncConfig>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub collections: Option<Vec<CollectionRef>>,
    #[serde(rename = "createdAt")]
    pub created_at: String,
    #[serde(
        default,
        rename = "lastOpenedAt",
        skip_serializing_if = "Option::is_none"
    )]
    pub last_opened_at: Option<String>,

    // ── Modules & settings ───────────────────────────────────────
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub modules: Option<IndexMap<String, bool>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub settings: Option<IndexMap<String, Value>>,

    // ── Ownership ────────────────────────────────────────────────
    #[serde(default, rename = "ownerId", skip_serializing_if = "Option::is_none")]
    pub owner_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub visibility: Option<ManifestVisibility>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,

    // ── Access control ───────────────────────────────────────────
    #[serde(
        default,
        rename = "privilegeSets",
        skip_serializing_if = "Option::is_none"
    )]
    pub privilege_sets: Option<Vec<PrivilegeSet>>,
    #[serde(
        default,
        rename = "roleAssignments",
        skip_serializing_if = "Option::is_none"
    )]
    pub role_assignments: Option<Vec<RoleAssignment>>,
}
